README.md
# Tailon [![Apache License](https://img.shields.io/badge/license-Apache-blue.svg)](https://github.com/NiNiyas/tailon/blob/master/LICENSE) [![GitHub release](https://img.shields.io/github/release/NiNiyas/tailon.svg)](https://github.com/NiNiyas/tailon/releases)

<p align="center">
  <img src="images/tailon.png">
</p>

Fork of [gvalkov/tailon](https://github.com/gvalkov/tailon).

Tailon is a webapp for looking at and searching through files and streams. \
In a nutshell, it is a fancy web wrapper around the following commands:

```
tail -f
tail -f | grep
tail -f | awk
tail -f | sed
```

What sets tailon apart from other similar projects is:

* Fully self-contained executable. Just download (or build) and run.
* Responsive and minimal user-interface.

## Install

### Docker
#### docker run

```
docker run --rm ghcr.io/niniyas/tailon:beta --help
```

#### docker compose
```
version: "3.9"
services:
  tailon:
    container_name: Tailon
    image: ghcr.io/niniyas/tailon:beta
    ports:
      - 8080:8080
    volumes:
      - /var/log/syslog:/syslog
      - /var/log/log1:/log1
      - .tailon/config:/config  # If you need to load from config file, put your config.toml in this folder.
    command: -b :8080 "group=Syslog,alias=syslog,/syslog" "group=group1,/log1/*.log"
    command: -c /config/config.toml # If you need to load from config file
```

### Releases

Download a build for your platform from the [releases](https://github.com/NiNiyas/tailon/releases) page.

## Changelog

See [changelog](CHANGELOG.md).

## Usage

Tailon is a command-line program that starts a local HTTP server, which in turn streams the output of commands such as `tail` and `grep`. \
It can be configured from its command-line interface or through the convenience of a [toml](https://github.com/toml-lang/toml) config file. \
Some options, like adding new commands, are only available through the configuration file.

To get started, run tailon with the list of files that you wish to monitor.

```
tailon /var/log/apache/access.log /var/log/apache/error.log /var/log/messages
```

Tailon can serve single files, globs or whole directory trees. \
Tailon’s server-side functionality is summarized entirely in its help message:

[//]: # (run "make README.md" to update the next section with the output of tailon --help)

[//]: # (BEGIN HELP)
```
Usage: tailon -c <config file>
Usage: tailon [options] <filespec> [<filespec> ...]

Tailon is a webapp for looking at and searching through files and streams.

  -a, --allow-download         Allow file downloads. (default true)
  -b, --bind string            Listen on the specified address and port (default ":8080")
  -c, --config string          Config.toml file location.
  -h, --help                   Show this help message and exit
  -e, --help-config            Show configuration file help and exit
      --history-lines int      No. of history lines to tail.
      --lines-to-tail int      No. of lines to tail. (default 100)
  -r, --relative-root string   Webapp relative root. (default "/")

Tailon can be configured through a config file or with command-line flags.

The command-line interface expects one or more filespec arguments, which
specify the files to be served. The expected format is:

  [alias=name,group=name]<spec>

where <spec> can be a file name, glob or directory. The optional 'alias='
and 'group=' specifiers change the display name of the files in the UI and
the group in which they appear.

A file specifier points to a single, possibly non-existent file. The file
name in the UI can be overwritten with 'alias='. For example:

  tailon alias=error.log,/var/log/apache/error.log

A glob evaluates to the list of files that match a shell file name pattern.
The pattern is evaluated each time the file list is refreshed. An 'alias='
specifier overwrites the parent directory of each matched file in the UI.

  tailon "/var/log/apache/*.log" "alias=nginx,/var/log/nginx/*.log"

If a directory is given, all files under it are served recursively.

  tailon /var/log/apache/ /var/log/nginx/

Example usage:
  tailon file1.txt file2.txt file3.txt
  tailon alias=messages,/var/log/messages "/var/log/*.log"
  tailon -b localhost:8080,localhost:8081 -c config.toml

For information on usage through the configuration file, please refer to the
'--help-config' option.
```
[//]: # (END HELP)

### Config File
Tailon can be configured through a TOML config file. \
The config file allows more configurability than the command-line interface.

```
  # The <title> of the index page.
  title = "Tailon"

  # The root of the web application.
  relative-root = "/"

  # The addresses to listen on. Can be an address:port combination or an unix socket.
  listen-addr = [":8080"]

  # Allow download of know files (only those matched by a filespec).
  allow-download = true

  # Commands that will appear in the UI.
  allow-commands = ["tail", "grep", "grep -v", "sed", "awk"]

  title = "Tailon"
  relative-root = "/"
  listen-addr = [":8080"]
  allow-download = true
  lines-of-history = 0
  lines-to-tail = 100
  allow-commands = ["tail", "grep", "grep -v", "sed", "awk"]

  [files]
    file1 = "alias=test,group=test_group,log.log"
    file2 = "alias=test1,group=test_group1,log.log"

  [commands]
    [commands.tail]
    action = ["tail", "-n", "$lines", "-F", "$path"]

    [commands.grep]
    stdin = "tail"
    action = ["grep", "-e", "$script"]
    default = ".*"

    [commands.sed]
    stdin = "tail"
    action = ["sed", "-u", "-e", "$script"]
    default = "s/.*/&/"

    [commands.awk]
    stdin = "tail"
    action = ["awk", "--sandbox", "$script"]
    default = "{print $0; fflush()}"

    [commands."grep -v"]
    stdin = "tail"
    action = ["grep", "-v", "--text", "--line-buffered", "--color=never", "-e", "$script"]
    default = "^$"
```

### Labels

Tailon will automatically convert the following to labels:

#### Log levels

- EMERGENCY
- ALERT
- CRITICAL
- ERROR
- WARNING
- WARN
- NOTICE
- INFO
- DEBUG

#### HTTP Methods

- GET
- POST
- PUT
- HEAD
- DELETE
- PATCH
- OPTIONS
- CONNECT
- TRACE

#### Others
- FAILED
- FAILURE

## Security

Tailon runs commands on the server it is installed on. While commands that
accept a script argument (such as awk, sed and grep) should be invulnerable to
shell injection, they may still allow for arbitrary command execution and
unrestricted access to the filesystem.

To clarify this point, consider the following input to the sed command:

```
s/a/b'; cat /etc/secrets
```

This will result in an error, as tailon does not invoke commands through a
shell. On the other hand, the following command is a perfectly valid sed script
that has the same effect as the above attempt for shell injection:

```
r /etc/secrets
```

The default set of enabled commands - tail, grep and awk - should be safe to
use. GNU awk is run in [sandbox](http://www.gnu.org/software/gawk/manual/html_node/Options.html#index-g_t_0040code_007b_002dS_007d-option-277) mode, which prevents scripts from accessing your
system, either through the `system()` builtin or by using input redirection.

By default, tailon is accessible to anyone who knows the server address and
port.


## Development

### Requirements

If you are on Windows, use WSL2. Install [Ubuntu](https://apps.microsoft.com/store/detail/ubuntu/9PDXGNCFSCZV) from Microsoft Store.

- GO: v1.18. I followed [this](https://www.digitalocean.com/community/tutorials/how-to-install-go-on-ubuntu-20-04) guide.
- Node.js: v16.*. I followed [this](https://github.com/nodesource/distributions/blob/master/README.md) guide.
- Make: `sudo apt-get install make`
- [entr](https://github.com/eradman/entr): `sudo apt-get install entr`


```
git clone https://github.com/NiNiyas/tailon.git
cd tailon
export GOROOT=/usr/local/go
export GOPATH=$HOME/go
export PATH=$PATH:$GOROOT/bin
go get
```

**Note**: The paths above might be different for you. Set this accordingly.

### Frontend

The web interface is a written in plain ES5 with the help of some Vue.js. A
simple makefile is used to bundle and compress the frontend assets. To work on
the frontend, make sure you're building with the `dev` build tag:

```
go build -tags dev
```

This will ensure that the `tailon` binary is reading assets from the
`frontend/dist` directory instead of from `frontend/assets_vfsdata.go`. To
compile the web assets, use `make all` or `make all BUILD=dev` in case you want
to simply concatenate files instead of also compressing them.

The `make watch` goal can be used to continuously update the bundles as you make
changes to the sources. This requires [entr](#requirements).

Note that the minified frontend bundles are committed in order to avoid people
wanting to work only on the backend from having to pull the full `node_modules`.

To build frontend, I have included a simple [build_frontend](https://github.com/NiNiyas/tailon/blob/master/build_frontend.sh) script. \
Make it executable with `sudo chmod +x build_frontend.sh` and run it `./build_frontend.sh`.

### Backend

The backend is written in straightforward go that tries to do as much as
possible using only the standard library.

#### Alpine Linux build

- INFO: http://krolow.com.br/til/go-compile-binary-not-running-in-alpine-and-busybox/

`CGO_ENABLED=0 go build -tags netgo -a -v`


## Backlog

See [TODO](TODO.md).

## Known Bugs

- When toolbar is hidden, last few lines are not visible in Fennec FDroid 101.1.1 build#1011120 8092b4e74+ and possibly other Firefox versions as well. Works fine on Bromite.

### Testing

The project has unit-tests, which you can run with `go test` and integration
tests which you can run with `cd tests; pytest`. \
Alternatively, you can run both with `make test`.

The integration tests are written in Python and use `pytest` and `aiohttp` to
interact with a running `tailon` instance. \
To run the integration tests, you first need to install the needed dependencies:

**Note**: Python testing is broken. I have commented out the line in [Makefile](Makefile), I will take a look at it when I have time.

```shell
# Option 1: Using a virtualenv
python3 -m venv path/to/venv
source path/to/venv/bin/activate
python3 -m pip install -r tests/requirements.txt

# Option 2: User site-packages
python3 -m pip install --user -r tests/requirements.txt
```

### Contribute

Please do contribute! Issues and pull requests are welcome. I could use some help.

<!--## What about the other tailon project?

This project is a full rewrite of the original [tailon](https://github.com/gvalkov/tailon-legacy) with the following goals in mind:

* Reduce maintenance overhead (especially on the frontend).
* Remove unwanted features and fix poor design choices.
* Learn more about Go and Vue.js.

In terms of tech, the following has changed:

* Backend from Python+Tornado to Go.
* Frontend from a very-custom Typescript solution to a simple ES5 + Vue.js app.
* Simplified asset pipeline (a short Makefile).
* Config file is now toml based.
* Fully self-contained executable.


## Similar Projects

* [clarity](https://github.com/tobi/clarity)
* [errorlog](http://www.psychogenic.com/en/products/Errorlog.php)
* [log.io](http://logio.org/)
* [rtail](http://rtail.org/)
* [tailon](https://github.com/gvalkov/tailon-legacy) -->


CHANGELOG.md
# Changelog

## [1.0.3] - 2022-12-13

## UI
- Added support for WARN label.

## Other
- Add ignored build files. Thanks to [sam0324sam](https://github.com/sam0324sam) for reporting this.

## [1.0.2] - 2022-07-21

## UI
- Another possible fix for default font issue (#1).
- Added `Gainsboro` font color.
- Changed default font-family of log-view to `Consolas, Menlo,Monaco, Lucida Console, Liberation Mono, DejaVu Sans Mono, Bitstream Vera Sans Mono, Courier New, monospace, serif`. Same as [gvalkov/tailon](https://github.com/gvalkov/tailon/blob/master/frontend/scss/vars.scss#L4)

## [1.0.1] - 2022-07-21

## UI
- Possible fix for default font issue (#1).
- Toolbar fonts will now be changed too.
- Reduced the default font size from 15px to 13px.

## [1.0.0] - 2022-07-06
First of all, I am not familiar with GO or JavaScript. I learned them while I was trying to make this better. \
So you may find some codes here and there that were not meant to be and I apologize for that.

## UI
- A bunch of CSS tweaks
- Wrapped lines mode will have alternative colors between lines.
- Configuration dialog will now be hidden when clicking elsewhere.
- The logs will now fit most of the screens properly.
- Added the option to change background color and font color. Thanks to [nurzico/tailon](https://github.com/nurzico/tailon).
- Added labels for log levels. Thanks to [fquinto/tailon](https://github.com/fquinto/tailon) for the code. I added the rest of them.
- Added labels for all the HTTP methods.
- Added the ability to clear search field with escape button. Thanks to [bformet/tailon](https://github.com/bformet/tailon/commit/03d08ab3b36c9b982c09754d93ecfb8fdbd54284)
- Favicon is now displaying properly.
- Added a separate configuration for UI changes.
- Added the option for changing font type.
- Download file option now opens in a new window.
- Changed toolbar icons. They were from:
  - [Eraser](https://www.svgrepo.com/svg/72555/eraser)
  - [Settings](https://www.svgrepo.com/svg/10074/settings)
  - [Download](https://www.svgrepo.com/svg/14689/download)
  - [UI Tools](https://www.reshot.com/free-svg-icons/item/tool-YANDWEVLTJ/)
- Added a range slider to change font size. It is from [W3Schools](https://www.w3schools.com/howto/howto_js_rangeslider.asp).

## CLI
- Added command line `lines-to-tail` and `lines-of-history`. Thanks to [ebbeknudsen/tailon](https://github.com/ebbeknudsen/tailon/commit/944862f466790dfe67715c8c82f6edbd5605915a)
- Formatted help message.

## Config File
- Added the ability to load files from config.toml. Thanks to [joshuaboniface/tailon](https://github.com/joshuaboniface/tailon/commit/8d1a949e563a719938c1d74be693e49d57e246ed).  \
You can view the example in [README](README.md#config-file) or [config.toml](config.toml). This code would only load multiple files from toml file but shows only the last one when using command line. I somehow managed to fix this. I am not sure how reliable it is. I just put one or two lines in 314-316 lines in main.go file. But in my testing it worked fine.

## Dependencies
- Updated frontend dependencies:
  - autoprefixer: 8.6.1 -> 10.4.7
  - clean-css: 4.1.11 -> 5.3.0
  - clean-css-cli: 4.1.11 -> 5.6.0
  - debounce: 1.1.0 -> 1.2.1
  - node-sass: 4.11.0 -> 7.0.1
  - npm: 6.13.4 -> 8.13.2
  - postcss-cli: 5.0.0 -> 10.0.0
  - sockjs-client: 1.3.0 -> 1.6.1
  - uglify-js: 3.4.4 -> 3.16.2
  - vue: 2.5.16 -> 2.7.2
  - vue-loading-overlay: 2.1.0 -> 3.4.2
  - vue-multiselect: 2.1.0 -> 2.1.6
- Added `@mdi/font` as a dependency.
- Updated updated materialdesignicons webfonts and css in `dist` folder from v5.8.55 -> v6.9.96.
- Updated sockjsv2 -> sockjsv3 in `server.go` file. Thanks to [GChuf/tailon](https://github.com/GChuf/tailon/commit/5684abaf7cae3722e1a659afcf0ff2ea8172a707).

## Development
- Added developement requirements and steps in  [README](README.md). Huge thanks to [fquinto/tailon](https://github.com/fquinto/tailon) for including how to build in their README.
- Added `license`, `description`, `scripts`, `keywords`, `bugs` and `homepage` fields in `package.json` file.
- Added `npm run` commands to build frontend and tailon.
   - `npm run build` will build dependencies, upgrade them, delete old .css and .js files and rebuilds them and exits.
   - `npm run build-and-watch` will also build dependencies, upgrade them, delete old .css and .js files and rebuilds them and watches for changes in frontend files. This requires [entr](https://github.com/eradman/entr).
   - `npm run build-tailon` will build tailon for your arch and exits.
   - `npm run build-frontend-and-tailon` will build dependencies, upgrade them, delete old .css and .js files, rebuilds them and also build tailon and exits.

## Docker
- Added Dockerfile healthcheck.
- Removed `upx` dependency.
- Added multi-arch docker images, built with GitHub Actions.

## Miscellaneous
- `make all` command in frontend directory will beautify all the .js files and .css files that it creates.
- Favicon location was moved from `frontend/dist` directory to `frontend/dist/static/favicon` folder.
- Added `grep -v` command. Thanks to [alexeykobzev/tailon](https://github.com/alexeykobzev/tailon/commit/034b17b66c08d8ab3430b6e1521821969ddb7d99).
- `make README.md` command in root folder is broken. I have commented it out.


server.go
package main

import (
	"encoding/json"
	"github.com/gorilla/handlers"
	"tailon/cmd"
	"tailon/frontend"
	"github.com/shurcooL/httpfs/html/vfstemplate"
	"github.com/shurcooL/httpgzip"
	"gopkg.in/igm/sockjs-go.v3/sockjs"
	"html/template"
	"log"
	"net/http"
	"os"
	"os/exec"
	"strconv"
	"time"
)

func setupRoutes(relativeroot string) *http.ServeMux {
	router := http.NewServeMux()

	// This is either "frontend/frontend.go" or "frontend/assets_vfsdata.go", depending
	// on the "dev" build tag. The latter is generated by "frontend/Makefile" with "vfsgen".
	staticHandler := noCacheControl(httpgzip.FileServer(frontend.Assets, httpgzip.FileServerOptions{IndexHTML: true}))

	sockjsHandler := sockjs.NewHandler(relativeroot+"ws", sockjs.DefaultOptions, wsHandler)
	staticHandler = http.StripPrefix(relativeroot+"vfs/", staticHandler)

	router.Handle(relativeroot+"vfs/", staticHandler)
	router.Handle(relativeroot+"ws/", sockjsHandler)
	router.HandleFunc(relativeroot+"files/", downloadHandler)
	router.HandleFunc(relativeroot+"", indexHandler)

	return router
}

func setupServer(config *Config, addr string, logger *log.Logger) *http.Server {
	router := setupRoutes(config.RelativeRoot)
	loggingRouter := handlers.LoggingHandler(os.Stderr, router)

	server := http.Server{
		Addr:         addr,
		Handler:      loggingRouter,
		ErrorLog:     logger,
		ReadTimeout:  5 * time.Second,
		WriteTimeout: 10 * time.Second,
		IdleTimeout:  15 * time.Second,
	}

	return &server
}

func indexHandler(w http.ResponseWriter, r *http.Request) {
	t := template.Must(vfstemplate.ParseFiles(frontend.Assets, nil, "/templates/base.html", "/templates/tailon.html"))
	t.Execute(w, config)
}

func downloadHandler(w http.ResponseWriter, r *http.Request) {
	if !config.AllowDownload {
		http.Error(w, "downloads forbidden by server", http.StatusForbidden)
		return
	}

	path := r.URL.Query()["path"][0]
	if !fileAllowed(path) {
		log.Printf("warn: attempt to access unknown file: %s", path)
		http.Error(w, "unknown file", http.StatusNotFound)
		return
	}
	http.ServeFile(w, r, path)
}

func noCacheControl(h http.Handler) http.Handler {
	fn := func(w http.ResponseWriter, r *http.Request) {
		w.Header().Set("Cache-Control", "no-store, no-cache, must-revalidate, max-age=0")
		h.ServeHTTP(w, r)
	}

	return http.HandlerFunc(fn)
}

// FrontendCommand instances are the messages that the client sends to the server when the file, tool or script change.
type FrontendCommand struct {
	Command string
	Script  string
	Entry   ListEntry
	Nlines  int
}

// The main sockjs handler.
func wsHandler(session sockjs.Session) {
	messages := make(chan string)
	done := make(chan struct{})
	defer close(done)

	go wsWriter(session, messages, done)

	for {
		if msg, err := session.Recv(); err == nil {
			messages <- msg
			continue
		} else {
			log.Print(err)
		}
		break
	}
}

// Goroutine handling received messages and streaming of file contents.
func wsWriter(session sockjs.Session, messages chan string, done <-chan struct{}) {
	// The processes that make up the pipeline. The stdout of procA is connected to  the stdin of procB.
	var procA *exec.Cmd
	var procB *cmd.Cmd

	cmdOptions := cmd.Options{Buffered: false, Streaming: true}

	for {
		select {
		case msg := <-messages:
			if msg == "list" {
				lst := createListing(config.FileSpecs)
				b, err := json.Marshal(lst)
				if err != nil {
					log.Println("error: ", err)
				}
				session.Send(string(b))
			} else if msg[0] == '{' {
				msgJSON := FrontendCommand{}
				json.Unmarshal([]byte(msg), &msgJSON)

				if !fileAllowed(msgJSON.Entry.Path) {
					log.Print("Unknown file: ", msgJSON.Entry.Path)
					continue
				}

				killProcs(procA, procB)

				// Check if the command is using another command for stdin.
				stdinSource := config.CommandSpecs[msgJSON.Command].Stdin
				if stdinSource != "" {
					actionA := config.CommandSpecs[stdinSource].Action
					actionA = expandCommandArgs(actionA, msgJSON)
					procA = exec.Command(actionA[0], actionA[1:]...)
					log.Print("Running command: ", actionA)
				}

				actionB := config.CommandSpecs[msgJSON.Command].Action
				actionB = expandCommandArgs(actionB, msgJSON)
				procB = cmd.NewCmdOptions(cmdOptions, actionB[0], actionB[1:]...)
				log.Print("Running command: ", actionB)

				// Start streaming procB's stdout and stderr to the client.
				go streamOutput(procA, procB, session)
			}
		case <-done:
			killProcs(procA, procB)
			return
		}
	}
}

// Expands the variables in main.CommandSpec.Action with the values in the
// frontend command. For example:
//    ["tail", "-n", "$lines", "-F", "$path"] -> ["tail", "-n", "10", "-F", "f1.txt"]
func expandCommandArgs(action []string, cmd FrontendCommand) []string {
	var res = make([]string, 0)

	for _, arg := range action {
		switch arg {
		case "$lines":
			res = append(res, strconv.Itoa(cmd.Nlines))
		case "$path":
			res = append(res, cmd.Entry.Path)
		case "$script":
			res = append(res, cmd.Script)
		default:
			res = append(res, arg)
		}
	}

	return res
}

// Goroutine that streams command stdout and stderr to the client.
func streamOutput(procA *exec.Cmd, procB *cmd.Cmd, session sockjs.Session) {
	if procA != nil {
		procB.Stdin, _ = procA.StdoutPipe()
		procA.Start()
	}

	statusChan := procB.Start()

	for {
		select {
		case line := <-procB.Stdout:
			msg := []string{"o", line}
			data, _ := json.Marshal(msg)
			session.Send(string(data))
		case line := <-procB.Stderr:
			msg := []string{"e", line}
			data, _ := json.Marshal(msg)
			session.Send(string(data))
		case <-statusChan:
		}
	}
}

func killProcs(procA *exec.Cmd, procB *cmd.Cmd) {
	if procA != nil {
		log.Printf("Stopping pid %d", procA.Process.Pid)
		procA.Process.Kill()
		procA.Wait()
	}

	if procB != nil {
		log.Printf("Stopping pid %d", procB.Status().PID)
		if procB.Stdin != nil {
			procB.Stdin.Close()
		}
		procB.Stop()
	}
}


frontend/templates/sw.js
const staticCacheName = 'site-static-v1';
const assets = [
  '/',
  '/index.html',
  '/assets/js/ui.js',
  '/assets/css/main.css',
  '/assets/images/background-home.jpg',
  'https://fonts.googleapis.com/css?family=Lato:300,400,700',
];
// install event
self.addEventListener('install', evt => {
  evt.waitUntil(
    caches.open(staticCacheName).then((cache) => {
      console.log('caching shell assets');
      cache.addAll(assets);
    })
  );
});
// activate event
self.addEventListener('activate', evt => {
  evt.waitUntil(
    caches.keys().then(keys => {
      return Promise.all(keys
        .filter(key => key !== staticCacheName)
        .map(key => caches.delete(key))
      );
    })
  );
});
// When we change the name we could have multiple cache, to avoid that we need to delet the old cache, so with this function we check the key that is our cache naming, if it is different from the actual naming we delete it, in this way we will always have only the last updated cache.
// fetch event
self.addEventListener('fetch', evt => {
  evt.respondWith(
    caches.match(evt.request).then(cacheRes => {
      return cacheRes || fetch(evt.request);
    })
  );
});


main.go
// A webapp for looking at and searching through files.
package main

import (
	"fmt"
	"io/ioutil"
	"log"
	"net"
	"os"
	"strings"
	"sync"

	"github.com/mitchellh/mapstructure"
	"github.com/pelletier/go-toml"
	flag "github.com/spf13/pflag"
)

const scriptDescription = `
Usage: tailon -c <config file>
Usage: tailon [options] <filespec> [<filespec> ...]

Tailon is a webapp for looking at and searching through files and streams.
`

const scriptEpilog = `
Tailon can be configured through a config file or with command-line flags.

The command-line interface expects one or more filespec arguments, which
specify the files to be served. The expected format is:

  [alias=name,group=name]<spec>

where <spec> can be a file name, glob or directory. The optional 'alias='
and 'group=' specifiers change the display name of the files in the UI and
the group in which they appear.

A file specifier points to a single, possibly non-existent file. The file
name in the UI can be overwritten with 'alias='. For example:

  tailon alias=error.log,/var/log/apache/error.log

A glob evaluates to the list of files that match a shell file name pattern.
The pattern is evaluated each time the file list is refreshed. An 'alias='
specifier overwrites the parent directory of each matched file in the UI.

  tailon "/var/log/apache/*.log" "alias=nginx,/var/log/nginx/*.log"

If a directory is given, all files under it are served recursively.

  tailon /var/log/apache/ /var/log/nginx/

Example usage:
  tailon file1.txt file2.txt file3.txt
  tailon alias=messages,/var/log/messages "/var/log/*.log"
  tailon -b localhost:8080,localhost:8081 -c config.toml

For information on usage through the configuration file, please refer to the
'--help-config' option.
`

const configFileHelp = `
Tailon can be configured through a TOML config file. The config file allows
more configurability than the command-line interface.

  # The <title> of the index page.
  title = "Tailon"

  # The root of the web application.
  relative-root = "/"

  # The addresses to listen on. Can be an address:port combination or an unix socket.
  listen-addr = [":8080"]

  # Allow download of know files (only those matched by a filespec).
  allow-download = true

  # Commands that will appear in the UI.
  allow-commands = ["tail", "grep", "grep -v", "sed", "awk"]
`

const defaultTomlConfig = `
  title = "Tailon"
  relative-root = "/"
  listen-addr = [":8080"]
  allow-download = true
  lines-of-history = 0
  lines-to-tail = 100
  allow-commands = ["tail", "grep", "grep -v", "sed", "awk"]

  [files]
    file1 = "alias=test,group=test_group,log.log"
    file2 = "alias=test1,group=test_group1,log.log"

  [commands]
    [commands.tail]
    action = ["tail", "-n", "$lines", "-F", "$path"]

    [commands.grep]
    stdin = "tail"
    action = ["grep", "-e", "$script"]
    default = ".*"

    [commands.sed]
    stdin = "tail"
    action = ["sed", "-u", "-e", "$script"]
    default = "s/.*/&/"

    [commands.awk]
    stdin = "tail"
    action = ["awk", "--sandbox", "$script"]
    default = "{print $0; fflush()}"

    [commands."grep -v"]
    stdin = "tail"
    action = ["grep", "-v", "--text", "--line-buffered", "--color=never", "-e", "$script"]
    default = "^$"
`

// CommandSpec defines a command that the server can execute.
type CommandSpec struct {
	Stdin   string
	Action  []string
	Default string
}

func parseTomlConfig(config string) (*toml.Tree, map[string]CommandSpec, []FileSpec) {
	cfg, err := toml.Load(config)
	if err != nil {
		log.Fatal("Error parsing config: ", err)
	}

	commands := make(map[string]CommandSpec)

	cfgCommands := cfg.Get("commands").(*toml.Tree).ToMap()
	for key, value := range cfgCommands {
		command := CommandSpec{}
		err := mapstructure.Decode(value, &command)
		if err != nil {
			log.Fatal(err)
		}
		commands[key] = command
	}

	cfgFiles := cfg.Get("files").(*toml.Tree).ToMap()
	filespecs := make([]FileSpec, 0)
	for _, value := range cfgFiles {
        if value == nil {
            continue
        }
        strvalue := fmt.Sprint(value)
		if file, err := parseFileSpec(strvalue); err != nil {
			fmt.Fprintf(os.Stderr, "Error parsing argument '%s': %s\n", strvalue, err)
			os.Exit(1)
		} else {
            filespecs = append(filespecs, file)
		}
	}

	return cfg, commands, filespecs
}

// FileSpec is an instance of a file to be monitored. These are mapped to
// os.Args or the [files] elements in the config file.
type FileSpec struct {
	Path  string
	Type  string
	Alias string
	Group string
}

// Parse a string into a filespec. Example inputs are:
//   alias=1,group=2,/var/log/messages
//   /var/log/
//   /var/log/*
func parseFileSpec(spec string) (FileSpec, error) {
	var filespec FileSpec
	var path string
	parts := strings.Split(spec, ",")

	if length := len(parts); length == 1 {
		path = parts[0]
	} else {
		// The last part is the path. We'll probably need a more robust
		// solution in the future.
		path, parts = parts[len(parts)-1], parts[:len(parts)-1]
	}

	if strings.ContainsAny(path, "*?[]") {
		filespec.Type = "glob"
	} else {
		stat, err := os.Lstat(path)
		if os.IsNotExist(err) || stat.Mode().IsRegular() {
			filespec.Type = "file"
		} else if stat.Mode().IsDir() {
			filespec.Type = "dir"
		}
	}

	for _, part := range parts {
		if strings.HasPrefix(part, "group=") {
			group := strings.SplitN(part, "=", 2)[1]
			group = strings.Trim(group, "'\" ")
			filespec.Group = group
		} else if strings.HasPrefix(part, "alias=") {
			filespec.Alias = strings.SplitN(part, "=", 2)[1]
		}
	}

	if filespec.Type == "" {
		filespec.Type = "file"
	}
	filespec.Path = path
	return filespec, nil

}

// Config contains all backend and frontend configuration options and relevant state.
type Config struct {
	RelativeRoot      string
	BindAddr          []string
	ConfigPath        string
	WrapLinesInitial  bool
	TailLinesInitial  int
	LinesToTail       int64
	LinesOfHistory    int64
	AllowCommandNames []string
	AllowDownload     bool

	CommandSpecs   map[string]CommandSpec
	CommandScripts map[string]string
	FileSpecs      []FileSpec
}

func makeConfig(configContent string) *Config {
	defaults, commandSpecs, fileSpecs := parseTomlConfig(configContent)

	// Convert the list of bind addresses from []interface{} to []string.
	addrsA := defaults.Get("listen-addr").([]interface{})
	addrsB := make([]string, len(addrsA))
	for i := range addrsA {
		addrsB[i] = addrsA[i].(string)
	}

	config := Config{
		BindAddr:       addrsB,
		RelativeRoot:   defaults.Get("relative-root").(string),
		LinesToTail:    defaults.Get("lines-to-tail").(int64),
		LinesOfHistory: defaults.Get("lines-of-history").(int64),
		AllowDownload:  defaults.Get("allow-download").(bool),
		CommandSpecs:   commandSpecs,
		FileSpecs:     fileSpecs,
	}

	mapstructure.Decode(defaults.Get("allow-commands"), &config.AllowCommandNames)
	return &config
}

var config = &Config{}

func main() {
	config = makeConfig(defaultTomlConfig)

	printHelp := flag.BoolP("help", "h", false, "Show this help message and exit")
	printConfigHelp := flag.BoolP("help-config", "e", false, "Show configuration file help and exit")
	bindAddr := flag.StringP("bind", "b", strings.Join(config.BindAddr, ","), "Listen on the specified address and port")

	flag.StringVarP(&config.RelativeRoot, "relative-root", "r", config.RelativeRoot, "Webapp relative root.")
	flag.BoolVarP(&config.AllowDownload, "allow-download", "a", config.AllowDownload, "Allow file downloads.")
	flag.Int64Var(&config.LinesToTail, "lines-to-tail", config.LinesToTail, "No. of lines to tail.")
	flag.Int64Var(&config.LinesOfHistory, "history-lines", config.LinesOfHistory, "No. of history lines to tail.")
	flag.StringVarP(&config.ConfigPath, "config", "c", "", "Config.toml file location.")
	flag.Parse()

	flag.Usage = func() {
		fmt.Fprintln(os.Stderr, strings.TrimLeft(scriptDescription, "\n"))
		flag.PrintDefaults()
		fmt.Fprintln(os.Stderr, strings.TrimRight(scriptEpilog, "\n"))
		os.Exit(0)
	}

	if *printHelp {
		flag.Usage()
		os.Exit(0)
	}

	if *printConfigHelp {
		fmt.Fprintf(os.Stderr, "%s\n\n%s\n", strings.Trim(configFileHelp, "\n"), strings.Trim(defaultTomlConfig, "\n"))
		os.Exit(0)
	}

	config.BindAddr = strings.Split(*bindAddr, ",")

	// If a configuration file is specified, read all options from it. This discards all options set on the command-line.
	if config.ConfigPath != "" {
		if b, err := ioutil.ReadFile(config.ConfigPath); err != nil {
			fmt.Fprintf(os.Stderr, "Error reading config file '%s': %s\n", config.ConfigPath, err)
			os.Exit(1)
		} else {
			config = makeConfig(string(b))
		}
	}

	// Ensure that relative root is always '/' or '/$arg/'.
	config.RelativeRoot = "/" + strings.TrimLeft(config.RelativeRoot, "/")
	config.RelativeRoot = strings.TrimRight(config.RelativeRoot, "/") + "/"

	// Handle command-line file specs
	filespecs := make([]FileSpec, len(flag.Args()))
	for _, spec := range flag.Args() {
		if filespec, err := parseFileSpec(spec); err != nil {
			fmt.Fprintf(os.Stderr, "Error parsing argument '%s': %s\n", spec, err)
			os.Exit(1)
		} else {
			config.FileSpecs = append(filespecs, filespec) // from joshuaboniface fork
			filespecs = append(filespecs, filespec) // original one
			config.FileSpecs = filespecs // this wasn't here. It was on 319
		}
	}
	//config.FileSpecs = filespecs was here

	if len(config.FileSpecs) == 0 {
		fmt.Fprintln(os.Stderr, "No files specified on command-line or in config file")
		os.Exit(2)
	}

	config.CommandScripts = make(map[string]string)
	for cmd, values := range config.CommandSpecs {
		config.CommandScripts[cmd] = values.Default
	}

	log.Print("Generating initial file listing...")
	createListing(config.FileSpecs)

	var wg sync.WaitGroup
	for _, addr := range config.BindAddr {
		wg.Add(1)
		go startServer(config, addr)
	}
	wg.Wait()

}

func startServer(config *Config, bindAddr string) {
	loggerHTML := log.New(os.Stdout, "", log.LstdFlags)
	loggerHTML.Printf("Server start, relative-root: %s, bind-addr: %s\n", config.RelativeRoot, bindAddr)

	server := setupServer(config, bindAddr, loggerHTML)

	if strings.Contains(bindAddr, ":") {
		server.ListenAndServe()
	} else {
		os.Remove(bindAddr)

		unixAddr, _ := net.ResolveUnixAddr("unix", bindAddr)
		unixListener, err := net.ListenUnix("unix", unixAddr)
		unixListener.SetUnlinkOnClose(true)

		if err != nil {
			panic(err)
		}

		defer unixListener.Close()
		server.Serve(unixListener)
	}
}
